src/storage/gpu_swap_history.cc

```diff
@@ -81,5 +105,33 @@ void MemHistory::StopIteration() {
   ++iteration_idx_;
 }
 
+MemHistory::MemRecord MemHistory::find(std::vector<MemHistory::MemRecord> v,
+    size_t step) {
+  size_t i = 0;
+  size_t j = v.size() - 1;
+  while(i < j) {
+    if(i == v.size()-1)
+      return v[i];
+    if(j == 0)
+      return v[j];
+    size_t mid = (i + j) / 2;
+    size_t c_step = v[mid].record_step;
+    bool right_before = c_step < step && v[mid+1].record_step > step;
+    if(c_step == step || right_before) {
+      return v[mid+1];
+    } else if(c_step < step) {
+      i = mid + 1;
+    } else{
+      j = mid - 1;
+    }
+  }
```

Comment:
For this binary search, since you are working with a vector, I suggest using std::upper_bound / lower_bound, which can simply the code a lot and prevent bug.